app/src/main/java/org/mozilla/reference/browser/test/OpenClass.kt

```diff
@@ -0,0 +1,20 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.mozilla.reference.browser.test
+
+/**
+ * Annotate a class with [OpenClass] to open a class for mocking purposes while keeping it final in release builds
+ */
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+annotation class OpenClass
+
+/**
+ * Annotate a class with [Mockable] to make it extensible in debug builds
+ */
+@OpenClass
+@Target(AnnotationTarget.CLASS)
+annotation class Mockable
```

Comment:
With Mockito, adding a file like this should be enough: https://github.com/mozilla-mobile/android-components/blob/master/components/lib/crash/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker

Reply:
It's a different semantic and different scope.

First of all, Fenix has the same approach: see the [depencies file](https://github.com/mozilla-mobile/fenix/blob/8296b3e9ffe5b0067a9551caf24513b185117306/buildSrc/src/main/java/Dependencies.kt#L184) and the [OpenClass.kt](https://github.com/mozilla-mobile/fenix/blob/8296b3e9ffe5b0067a9551caf24513b185117306/app/src/main/java/org/mozilla/fenix/test/OpenClass.kt) (I just copied it from Fenix).

Because robolectric, we need to extend the BrowserApplication. There is no way in which we can avoid that, or at least no one I know about. Mocking doesn't help here:

1. by default, robolectric picks up, if it exists, a class named as the application class from the manifest prefixed with the word Test (so BrowserApplication -> TestBrowserApplication), not a specific instance on which I can mock some method.
2. we can configure which class to use also with annotations, but again is just the class not a specific instance.

I hope to have been clear enough in my explanation.